File: src/main/kotlin/com/example/controllers/maps/Maps.kt
package com.example.controllers.maps

import com.example.model.Snake
import com.example.utils.ELEMENT_SIZE
import com.example.utils.FIELD_HEIGHT
import com.example.utils.FIELD_WIDTH
import com.example.utils.divideCords

interface Map {
    val edges: MutableList<Pair<Int, Int>>
    val color: String
    fun isIntersected(snake: Snake): Boolean
}

abstract class AbstractMap(val name: String) : Map {
    override val edges: MutableList<Pair<Int, Int>> = ArrayList()
    override val color = "000000"

    override fun isIntersected(snake: Snake): Boolean {
        val headCords = snake.getHead().getCords()
        for (point in edges) {
            if (point == headCords)
                return true
        }
        return false
    }
}

class FreeMap: AbstractMap("free") {
    override fun isIntersected(snake: Snake): Boolean = false
}

class EdgesMap: AbstractMap("edges") {

    init {
        for (i in 0 until FIELD_WIDTH step ELEMENT_SIZE) {
            edges.add(Pair(i, 0))
            edges.add(Pair(i, FIELD_HEIGHT - ELEMENT_SIZE))
        }
        for (i in ELEMENT_SIZE until FIELD_HEIGHT - ELEMENT_SIZE step ELEMENT_SIZE) {
            edges.add(Pair(0, i))
            edges.add(Pair(FIELD_WIDTH - ELEMENT_SIZE, i))
        }

    }
}

class TunnelMap: AbstractMap("tunnel") {
    init {
        for (i in 0 until FIELD_HEIGHT / 4 step ELEMENT_SIZE) {
            edges.add(Pair(i, 0))
            edges.add(Pair(i, FIELD_HEIGHT - ELEMENT_SIZE))
            edges.add(Pair(FIELD_WIDTH - i - ELEMENT_SIZE, 0))
            edges.add(Pair(FIELD_WIDTH - i - ELEMENT_SIZE, FIELD_HEIGHT - ELEMENT_SIZE))
        }
        for (i in ELEMENT_SIZE until FIELD_HEIGHT / 4 - ELEMENT_SIZE step ELEMENT_SIZE) {
            edges.add(Pair(0, i))
            edges.add(Pair(FIELD_WIDTH - ELEMENT_SIZE, i))
            edges.add(Pair(0, FIELD_HEIGHT - ELEMENT_SIZE - i))
            edges.add(Pair(FIELD_WIDTH - ELEMENT_SIZE, FIELD_HEIGHT - i - ELEMENT_SIZE))
        }
        val temp = divideCords(FIELD_HEIGHT, 4)
        for (i in temp until FIELD_WIDTH - temp - ELEMENT_SIZE step ELEMENT_SIZE) {
            edges.add(Pair(i, temp))
            edges.add(Pair(i, FIELD_HEIGHT - temp))
        }
    }
}

class ApartmentMap: AbstractMap("apartment") {
    init {
        for (i in 0 until FIELD_WIDTH - ELEMENT_SIZE step ELEMENT_SIZE) {
            edges.add(Pair(i, FIELD_HEIGHT - divideCords(FIELD_HEIGHT, 3)))
        }
        for (i in 0 until FIELD_WIDTH / 2 - 5 * ELEMENT_SIZE step ELEMENT_SIZE) {
            edges.add(Pair(i, divideCords(FIELD_HEIGHT, 3)))
            edges.add(Pair(FIELD_WIDTH - i -  ELEMENT_SIZE, divideCords(FIELD_HEIGHT, 3)))
        }
        for (i in 0 until 20 * ELEMENT_SIZE step ELEMENT_SIZE) {
            edges.add(Pair(i, 0))
        }
        for (i in 35 * ELEMENT_SIZE until FIELD_WIDTH - 20 * ELEMENT_SIZE step ELEMENT_SIZE) {
            edges.add(Pair(i, 0))
        }
        for (i in ELEMENT_SIZE until ELEMENT_SIZE * 15 step ELEMENT_SIZE) {
            edges.add(Pair(0, i))
        }
        for (i in ELEMENT_SIZE until FIELD_HEIGHT / 3 step ELEMENT_SIZE) {
            edges.add(Pair(divideCords(FIELD_WIDTH, 2) - 5 * ELEMENT_SIZE, i))
        }
        for (i in FIELD_HEIGHT - divideCords(FIELD_HEIGHT, 3) + ELEMENT_SIZE until FIELD_HEIGHT step ELEMENT_SIZE) {
            edges.add(Pair(FIELD_WIDTH - divideCords(FIELD_WIDTH, 2) + 4 * ELEMENT_SIZE, i))
        }
    }
}

File: src/main/resources/webapp/js/maps.js
class FreeMap {

    constructor() {
        this.name = 'free';
        this.edges = [];
        this.color = 'black';
    }

    isIntersectedBy(snake) {
        return false;
    }

}

class EdgesMap {

    constructor() {
        this.name = 'edges'
        this.edges = [];
        this.color = 'black';

        for (let i = 0; i < FIELD_WIDTH; i += ELEMENT_SIZE) {
            this.edges.push({
                x: i,
                y: 0,
            })
            this.edges.push({
                x: i,
                y: FIELD_HEIGHT - ELEMENT_SIZE
            })
        }

        for (let i = ELEMENT_SIZE; i < FIELD_HEIGHT - ELEMENT_SIZE; i += ELEMENT_SIZE) {
            this.edges.push({
                x: 0,
                y: i
            })
            this.edges.push({
                x: FIELD_WIDTH - ELEMENT_SIZE,
                y: i
            })
        }
    }



    isIntersectedBy(snake) {
        let headCords = snake.getHead().getCords();
        for (let point of this.edges) {
            if (equals(headCords, point)) 
                return true;
        }
        return false;
    }

}

class TunnelMap {
    constructor() {
        this.name = 'tunnel';
        this.edges = [];
        this.color = 'black';

        for (let i = 0; i < FIELD_HEIGHT / 4; i += ELEMENT_SIZE) {
            this.edges.push({
                x: i,
                y: 0
            });
            this.edges.push({
                x: i,
                y: FIELD_HEIGHT - ELEMENT_SIZE
            });
            this.edges.push({
                x: FIELD_WIDTH - i - ELEMENT_SIZE,
                y: 0
            });
            this.edges.push({
                x: FIELD_WIDTH - i - ELEMENT_SIZE,
                y: FIELD_HEIGHT - ELEMENT_SIZE
            });
        }

        for (let i = ELEMENT_SIZE; i < FIELD_HEIGHT / 4 - ELEMENT_SIZE; i+=ELEMENT_SIZE) {
            this.edges.push({
                x: 0,
                y: i
            });
            this.edges.push({
                x: FIELD_WIDTH - ELEMENT_SIZE,
                y: i
            });
            this.edges.push({
                x: 0,
                y: FIELD_HEIGHT - ELEMENT_SIZE - i
            });
            this.edges.push({
                x: FIELD_WIDTH - ELEMENT_SIZE,
                y: FIELD_HEIGHT - i - ELEMENT_SIZE
            })
        }

        for(let i = FIELD_HEIGHT / 4; i < FIELD_WIDTH - FIELD_HEIGHT / 4 - ELEMENT_SIZE; i += ELEMENT_SIZE ) {
            this.edges.push({
                x: i,
                y: FIELD_HEIGHT / 4
            });
            this.edges.push({
                x: i,
                y: FIELD_HEIGHT - FIELD_HEIGHT / 4
            })
        }
    }

    isIntersectedBy(snake) {
        let headCords = snake.getHead().getCords();
        for (let point of this.edges) {
            if (equals(headCords, point)) 
                return true;
        }
        return false;
    }

}

class ApartmentMap {
    constructor() {
        this.name = 'tunnel';
        this.edges = [];
        this.color = 'black';

        for (let i = 0; i < FIELD_WIDTH - ELEMENT_SIZE; i += ELEMENT_SIZE) {
            this.edges.push({
                x: i,
                y: FIELD_HEIGHT - Math.floor(FIELD_HEIGHT / 3) + Math.floor(FIELD_HEIGHT / 3) % ELEMENT_SIZE
            });
        }

        for (let i = 0; i < FIELD_WIDTH / 2 - 5 * ELEMENT_SIZE; i+= ELEMENT_SIZE) {
            this.edges.push({
                x: i,
                y: Math.floor(FIELD_HEIGHT / 3) - Math.floor(FIELD_HEIGHT / 3) % ELEMENT_SIZE
            });
            this.edges.push({
                x: FIELD_WIDTH - i - ELEMENT_SIZE,
                y: Math.floor(FIELD_HEIGHT / 3) - Math.floor(FIELD_HEIGHT / 3) % ELEMENT_SIZE
            });
        }

        for (let i = 0; i < 20 * ELEMENT_SIZE; i+= ELEMENT_SIZE) {
            this.edges.push({
                x: i,
                y: 0
            });
        }

        for (let i = 35 * ELEMENT_SIZE; i < FIELD_WIDTH - 20 * ELEMENT_SIZE; i += ELEMENT_SIZE) {
            this.edges.push({
                x: i,
                y: 0
            });
        }

        for (let i = ELEMENT_SIZE; i < ELEMENT_SIZE * 15; i += ELEMENT_SIZE) {
            this.edges.push({
                x: 0,
                y: i
            });
        }

        for (let i = ELEMENT_SIZE; i < FIELD_HEIGHT / 3; i += ELEMENT_SIZE) {
            this.edges.push({
                x: FIELD_WIDTH / 2 - 5 * ELEMENT_SIZE,
                y: i
            });
        }

        for (let i = FIELD_HEIGHT - Math.floor(FIELD_HEIGHT / 3) + Math.floor(FIELD_HEIGHT / 3) % ELEMENT_SIZE + ELEMENT_SIZE; i < FIELD_HEIGHT; i += ELEMENT_SIZE) {
            this.edges.push({
                x: FIELD_WIDTH - FIELD_WIDTH / 2 + 4 * ELEMENT_SIZE,
                y: i
            });
        }
    }

    isIntersectedBy(snake) {
        let headCords = snake.getHead().getCords();
        for (let point of this.edges) {
            if (equals(headCords, point)) 
                return true;
        }
        return false;
    }
}